Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeSystem.cs

```diff
@@ -0,0 +1,304 @@
+            }
+        }
+
+        if (HasComp<CartridgeLoaderComponent>(cartComponent.LoaderUid))
+            UpdateUiState(uid, cartComponent.LoaderUid.Value, component);
+
+    }
+
+    /// <summary>
+    /// Sends the user's name to the server cache.
+    /// </summary>
+    private void SendName(EntityUid uid, MessagesCartridgeComponent component, CartridgeComponent cartComponent, string? address)
+    {
+        string name = GetUserName(cartComponent);
+
+        var packet = new NetworkPayload()
+        {
+            ["UserId"] = GetUserUid(cartComponent),
+            ["NewName"] = GetUserName(cartComponent),
+        };
+        _deviceNetworkSystem.QueuePacket(uid, address, packet);
+    }
+
+    /// <summary>
+    /// Retrieves the name of the given user from the last contacted server
+    /// </summary>
+    private string GetName(int key, MessagesCartridgeComponent component)
+    {
+        if (component.LastServer == null)
+            return Loc.GetString("messages-pda-connection-error");
+        return _messagesServerSystem.GetNameFromDict(component.LastServer, key);
+    }
+
+    /// <summary>
+    /// Returns the user's id in the messages system
+    /// </summary>
+    public int? GetUserUid(CartridgeComponent component)
+    {
+        var idComponent = GetIdCard(component);
+        if (idComponent == null)
+            return null;
+        return idComponent.MessagesId;
+    }
+
+    /// <summary>
+    /// Returns the user's name and job title
+    /// </summary>
+    public string GetUserName(CartridgeComponent component)
+    {
+        var idComponent = GetIdCard(component);
+        string job;
+        string name;
+        if (idComponent == null)
+            return Loc.GetString("messages-pda-unknown-name");
+        if (idComponent.FullName != null)
+        {
+            name = idComponent.FullName;
+        }
+        else
+        {
+            name = Loc.GetString("messages-pda-unknown-name");
+        }
+        if (idComponent.JobTitle != null)
+        {
+            job = idComponent.JobTitle;
+        }
+        else
+        {
+            job = Loc.GetString("messages-pda-unknown-job");
+        }
+        return $"{name} ({job})";
+    }
+
+    /// <summary>
+    /// Finds the id card in the PDA if present
+    /// </summary>
+    private IdCardComponent? GetIdCard(CartridgeComponent component)
+    {
+        var loaderUid = component.LoaderUid;
+        if (loaderUid == null ||
+        !TryComp(loaderUid.Value, out PdaComponent? pdaComponent))
+            return null;
+
+        return CompOrNull<IdCardComponent>(pdaComponent.ContainedId);
+
+    }
+
+    ///<summary>
+    ///Updates the ui state of a given cartridge
+    ///</summary>
+    public void ForceUpdate(EntityUid uid, MessagesCartridgeComponent component)
+    {
+        if (HasComp<CartridgeLoaderComponent>(Transform(uid).ParentUid))
+            UpdateUiState(uid, Transform(uid).ParentUid, component);
+    }
+
```

Comment:
Resolving components is expensive so please don't just put it in twice instead of putting it in a local variable.

You also get the component in a completely different way in OnPacketReceived instead of unifying it which is going to cause problems down the line.

Also getting the cartridge loader from an entity should be on cartridgeloadersystem not here.